### PR TITLE
SIP2-134: Add /admin/health to edge-sip2

### PR DESCRIPTION
**Pull request by pavankumar181**
## Purpose
[SIP2-134](https://issues.folio.org/browse/SIP2-134) Add /admin/health to edge-sip2

## Approach
Added Health check endpoint "/admin/health" and test cases to the edge-sip2 module.
Please find the Evidence below

image

## Pre-Merge Checklist:
 Before merging this PR, please go through the following list and take appropriate actions.

 - Does this PR meet or exceed the expected quality standards?
   - [ ] Code coverage on new code is 80% or greater
   - [ ] Duplications on new code is 3% or less
   - [ ] There are no major code smells or security issues
 - Does this introduce breaking changes?
   - [ ] Were any API paths or methods changed, added or removed?
   - [ ] Were there any schema changes?
   - [ ] Did any of the interface versions change?
   - [ ] Were permissions changed, added, or removed?
   - [ ] Are there new interface dependencies?
   - [ ] There are no breaking changes in this PR.
   - [ ] Check logging.

 If there are breaking changes, please **STOP** and consider the following:

 - What other modules will these changes impact?
 - Do JIRAs exist to update the impacted modules?
   - [ ] If not, please create them
   - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
   - [ ] Do they have all they appropriate links to blocked/related issues?
 - Are the JIRAs under active development?
   - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
 - Do PRs exist for these changes?
   - [ ] If so, have they been approved?

 Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.

 While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/main/java/org/folio/edge/sip2/MainVerticle.java [hunk 1]
```diff
@@ -82,6 +86,28 @@ private String getSanitizedConfig() {
   public void start(Promise<Void> startFuture) {
     log.debug("Startup configuration: {}", this::getSanitizedConfig);
 
+    HttpServer httpServer = vertx.createHttpServer();
+
+    httpServer.requestHandler(request -> {
+      log.info("path : {}",request.path());
```
Comment (gurleenkaurbp) [inline src/main/java/org/folio/edge/sip2/MainVerticle.java:92]: This log message has to be improved and the type should be debug. The frequency of the admin/health may bulge up the logs.

File: src/main/java/org/folio/edge/sip2/MainVerticle.java [hunk 2]
```diff
@@ -82,6 +86,28 @@ private String getSanitizedConfig() {
...
+      if (request.path().equals("/admin/health")) {
```
Comment (gurleenkaurbp) [inline src/main/java/org/folio/edge/sip2/MainVerticle.java:93]: Please add a constant for "/admin/health"

File: src/main/java/org/folio/edge/sip2/MainVerticle.java [hunk 3]
```diff
@@ -82,6 +86,28 @@ private String getSanitizedConfig() {
...
+        HttpServerResponse response = request.response();
+        response.setStatusCode(200);
+        response.putHeader("Content-Type", "text/plain");
+        response.end("OK");
+        log.info("statusCode : {}",response.getStatusCode());
```
Comment (gurleenkaurbp) [inline src/main/java/org/folio/edge/sip2/MainVerticle.java:98]: Same comment for the log as put at line number 92

File: src/main/java/org/folio/edge/sip2/MainVerticle.java [hunk 4]
```diff
@@ -82,6 +86,28 @@ private String getSanitizedConfig() {
...
+    });
+
+    httpServer.listen(HEALTH_CHECK_PORT, res -> {
+      if (res.succeeded()) {
+        log.info("Health endpoint is now listening!");
+      } else {
+        log.info("Failed!");
+      }
+    });
+
```
Comment (gurleenkaurbp) [inline src/main/java/org/folio/edge/sip2/MainVerticle.java:110]: I would have preferred this code be put in a dedicated method or a function. Also, when the health check fails should it be logged as info failed ? Can you please check the other modules how its being handled.

**Review by gurleenkaurbp — CHANGES_REQUESTED** [review]

File: src/main/java/org/folio/edge/sip2/MainVerticle.java (hunk 3 above)
Comment (steveellis) [inline src/main/java/org/folio/edge/sip2/MainVerticle.java:98]: Need space after the comma with your params.

File: src/main/java/org/folio/edge/sip2/MainVerticle.java [hunk 5]
```diff
@@ -82,6 +86,28 @@ private String getSanitizedConfig() {
...
+    httpServer.requestHandler(request -> {
+      log.info("path : {}",request.path());
+      if (request.path().equals("/admin/health")) {
+        HttpServerResponse response = request.response();
+        response.setStatusCode(200);
+        response.putHeader("Content-Type", "text/plain");
+        response.end("OK");
+        log.info("statusCode : {}",response.getStatusCode());
+        log.info("message : {}",response.getStatusMessage());
```
Comment (steveellis) [inline src/main/java/org/folio/edge/sip2/MainVerticle.java:99]: Space after the comma

File: src/main/java/org/folio/edge/sip2/MainVerticle.java [hunk 6]
```diff
@@ -82,6 +86,28 @@ private String getSanitizedConfig() {
...
+      }
+    });
+
+    httpServer.listen(HEALTH_CHECK_PORT, res -> {
+      if (res.succeeded()) {
+        log.info("Health endpoint is now listening!");
+      } else {
+        log.info("Failed!");
```
Comment (steveellis) [inline src/main/java/org/folio/edge/sip2/MainVerticle.java:107]: Exclamation points in log messages don't add much value. ;) Also "Failed" isn't a good log message. Imagine yourself trying to figure out what is happening when all you have is the log message "failed". What failed? And if something failed it's probably better to `log.error`.

Comment (pavankumar181) [reply src/main/java/org/folio/edge/sip2/MainVerticle.java]: Added log.error

File: src/main/java/org/folio/edge/sip2/MainVerticle.java (hunk 3 above)
Comment (pavankumar181) [reply src/main/java/org/folio/edge/sip2/MainVerticle.java]: Added

File: src/main/java/org/folio/edge/sip2/MainVerticle.java (hunk 2 above)
Comment (pavankumar181) [reply src/main/java/org/folio/edge/sip2/MainVerticle.java]: Constant added.

File: src/main/java/org/folio/edge/sip2/MainVerticle.java [hunk 7]
```diff
@@ -258,6 +264,30 @@ public void start(Promise<Void> startFuture) {
...
+
+    httpServer.requestHandler(request -> {
+      log.debug("path : {}", request.path());
+      if (request.path().equals(HEALTH_CHECK_PATH)) {
+        HttpServerResponse response = request.response();
+        response.setStatusCode(200);
+        response.putHeader("Content-Type", "text/plain");
+        response.end("OK");
+        log.debug("statusCode : {}", response.getStatusCode());
+        log.info("message : {}", response.getStatusMessage());
```
Comment (gurleenkaurbp) [inline src/main/java/org/folio/edge/sip2/MainVerticle.java:278]: This log message should be changed to a more informative message like "admin health check service response message : {}

**Review by gurleenkaurbp — CHANGES_REQUESTED** [review]

File: src/main/java/org/folio/edge/sip2/MainVerticle.java [hunk 8]
```diff
@@ -258,6 +264,30 @@ public void start(Promise<Void> startFuture) {
...
+      }
+    });
+
+    httpServer.listen(HEALTH_CHECK_PORT, res -> {
+      if (res.succeeded()) {
+        log.info("Health endpoint is now listening!");
+      } else {
+        log.error("Failed to start the health endpoint. Cause : {}", res.cause().getMessage());
```
[the remark this answers is not in the document]
Comment (pavankumar181) [reply src/main/java/org/folio/edge/sip2/MainVerticle.java]: Changed the logger message.

File: src/main/java/org/folio/edge/sip2/MainVerticle.java [hunk 9]
```diff
@@ -258,6 +264,30 @@ public void start(Promise<Void> startFuture) {
 
   }
 
+  private void startHealthCheckService() {
```
[the remark this answers is not in the document]
Comment (pavankumar181) [reply src/main/java/org/folio/edge/sip2/MainVerticle.java]: method name changed to callAdminHealthCheckService.

File: src/main/java/org/folio/edge/sip2/MainVerticle.java (hunk 7 above)
Comment (pavankumar181) [reply src/main/java/org/folio/edge/sip2/MainVerticle.java]: Changed the logger message.

**Review by gurleenkaurbp — APPROVED** [review]
Thanks @pavankumar181

**Comment by pavankumar181** [comment]
> Thanks @pavankumar181

Thanks for reviewing @gurleenkaurbp.

File: src/test/java/org/folio/edge/sip2/api/MainVerticleTests.java [hunk 10]
```diff
@@ -207,4 +212,18 @@ private void validateExpectedACSStatus(String acsResponse) {
...
+      Vertx vertx, VertxTestContext testContext) {
+
+    HttpClient client = vertx.createHttpClient();
+    client.request(HttpMethod.GET, 8081, "localhost", "/admin/health")
+        .compose(HttpClientRequest::send)
+        .compose(HttpClientResponse::body).onSuccess(body -> {
+          assertEquals("OK", body.toString());
+          testContext.completeNow();
+        });
+  }
```
[the remark this answers is not in the document]
Comment (pavankumar181) [reply src/test/java/org/folio/edge/sip2/api/MainVerticleTests.java:245]: Added test case for which returns 404 error.

File: src/test/java/org/folio/edge/sip2/api/MainVerticleTests.java [hunk 11]
```diff
@@ -207,4 +212,18 @@ private void validateExpectedACSStatus(String acsResponse) {
...
+  @Test
+  @DisplayName("Test /admin/health endpoint")
+  public void testHealthCheck(
+      Vertx vertx, VertxTestContext testContext) {
+
+    HttpClient client = vertx.createHttpClient();
+    client.request(HttpMethod.GET, 8081, "localhost", "/admin/health")
+        .compose(HttpClientRequest::send)
+        .compose(HttpClientResponse::body).onSuccess(body -> {
+          assertEquals("OK", body.toString());
```
[the remark this answers is not in the document]
Comment (pavankumar181) [reply src/test/java/org/folio/edge/sip2/api/MainVerticleTests.java]: done

File: src/main/java/org/folio/edge/sip2/MainVerticle.java [hunk 12]
```diff
@@ -258,6 +264,31 @@ public void start(Promise<Void> startFuture) {
...
+    httpServer.requestHandler(request -> {
+      log.debug("path : {}", request.path());
+      if (request.path().equals(HEALTH_CHECK_PATH)) {
+        HttpServerResponse response = request.response();
+        response.setStatusCode(200);
+        response.putHeader("Content-Type", "text/plain");
+        response.end("OK");
+        log.debug("statusCode : {}", response.getStatusCode());
+        log.info("Admin health check service response message : {}", response.getStatusMessage());
+      }
```
[the remark this answers is not in the document]
Comment (pavankumar181) [reply src/main/java/org/folio/edge/sip2/MainVerticle.java]: Added 404 status code in else part.

File: src/main/java/org/folio/edge/sip2/MainVerticle.java [hunk 13]
```diff
@@ -258,6 +264,31 @@ public void start(Promise<Void> startFuture) {
...
+    HttpServer httpServer = vertx.createHttpServer();
+
+    httpServer.requestHandler(request -> {
+      log.debug("path : {}", request.path());
+      if (request.path().equals(HEALTH_CHECK_PATH)) {
+        HttpServerResponse response = request.response();
+        response.setStatusCode(200);
+        response.putHeader("Content-Type", "text/plain");
+        response.end("OK");
+        log.debug("statusCode : {}", response.getStatusCode());
```
[the remark this answers is not in the document]
Comment (pavankumar181) [reply src/main/java/org/folio/edge/sip2/MainVerticle.java]: removed

File: src/test/java/org/folio/edge/sip2/api/MainVerticleTests.java [hunk 14]
```diff
@@ -207,4 +212,18 @@ private void validateExpectedACSStatus(String acsResponse) {
...
+
+  @Test
+  @DisplayName("Test /admin/health endpoint")
+  public void testHealthCheck(
+      Vertx vertx, VertxTestContext testContext) {
+
+    HttpClient client = vertx.createHttpClient();
+    client.request(HttpMethod.GET, 8081, "localhost", "/admin/health")
+        .compose(HttpClientRequest::send)
+        .compose(HttpClientResponse::body).onSuccess(body -> {
```
[the remark this answers is not in the document]
Comment (pavankumar181) [reply src/test/java/org/folio/edge/sip2/api/MainVerticleTests.java]: replaced with .onComplete(testContext.succeeding(body ->

File: src/main/java/org/folio/edge/sip2/MainVerticle.java [hunk 15]
```diff
@@ -258,6 +264,31 @@ public void start(Promise<Void> startFuture) {
...
+        log.info("Admin health check service response message : {}", response.getStatusMessage());
+      }
+    });
+
+    httpServer.listen(HEALTH_CHECK_PORT, res -> {
+      if (res.succeeded()) {
+        log.info("Health endpoint is now listening!");
+      } else {
+        log.error("The call to admin health check service failed due to : {}",
+            res.cause().getMessage());
```
[the remark this answers is not in the document]
Comment (pavankumar181) [reply src/main/java/org/folio/edge/sip2/MainVerticle.java]: Added exception

File: src/main/java/org/folio/edge/sip2/MainVerticle.java [hunk 16]
```diff
@@ -258,6 +264,31 @@ public void start(Promise<Void> startFuture) {
...
+      }
+    });
```
[the remark this answers is not in the document]
Comment (pavankumar181) [reply src/main/java/org/folio/edge/sip2/MainVerticle.java:281]: changed

File: src/main/java/org/folio/edge/sip2/MainVerticle.java [hunk 17]
```diff
@@ -258,6 +264,31 @@ public void start(Promise<Void> startFuture) {
...
+        response.putHeader("Content-Type", "text/plain");
+        response.end("OK");
+        log.debug("statusCode : {}", response.getStatusCode());
+        log.info("Admin health check service response message : {}", response.getStatusMessage());
+      }
+    });
+
+    httpServer.listen(HEALTH_CHECK_PORT, res -> {
+      if (res.succeeded()) {
+        log.info("Health endpoint is now listening!");
```
[the remark this answers is not in the document]
Comment (pavankumar181) [reply src/main/java/org/folio/edge/sip2/MainVerticle.java]: no

**Comment by pavankumar181** [comment]
> Looks good, thanks!

Thank you.

**Pull request merged**